src/optimagic/optimizers/pyensmallen_optimizers.py

```diff
@@ -0,0 +1,88 @@
+"""Implement ensmallen optimizers."""
+
+from dataclasses import dataclass
+
+import numpy as np
+import pyensmallen as pye
+from numpy.typing import NDArray
+
+from optimagic import mark
+from optimagic.optimization.algo_options import (
+    CONVERGENCE_FTOL_REL,
+    CONVERGENCE_GTOL_ABS,
+    MAX_LINE_SEARCH_STEPS,
+    STOPPING_MAXITER,
+)
+from optimagic.optimization.algorithm import Algorithm, InternalOptimizeResult
+from optimagic.optimization.internal_optimization_problem import (
+    InternalOptimizationProblem,
+)
+from optimagic.typing import AggregationLevel, NonNegativeFloat, PositiveInt
+
+LIMITED_MEMORY_MAX_HISTORY = 10
+"""Number of memory points to be stored (default 10)."""
+MIN_LINE_SEARCH_STEPS = 1e-20
+"""The minimum step of the line search."""
+MAX_LINE_SEARCH_TRIALS = 50
+"""The maximum number of trials for the line search (before giving up)."""
+ARMIJO_CONSTANT = 1e-4
+"""Controls the accuracy of the line search routine for determining the Armijo
+condition."""
+WOLFE_CONDITION = 0.9
+"""Parameter for detecting the Wolfe condition."""
+
+
+@mark.minimizer(
+    name="ensmallen_lbfgs",
+    solver_type=AggregationLevel.SCALAR,
+    is_available=True,
+    is_global=False,
+    needs_jac=True,
+    needs_hess=False,
+    supports_parallelism=False,
+    supports_bounds=False,
+    supports_linear_constraints=False,
+    supports_nonlinear_constraints=False,
+    disable_history=False,
+)
+@dataclass(frozen=True)
+class EnsmallenLBFGS(Algorithm):
+    limited_memory_max_history: PositiveInt = LIMITED_MEMORY_MAX_HISTORY
+    stopping_maxiter: PositiveInt = STOPPING_MAXITER
+    armijo_constant: NonNegativeFloat = ARMIJO_CONSTANT  # needs review
+    wolfe_condition: NonNegativeFloat = WOLFE_CONDITION  # needs review
+    convergence_gtol_abs: NonNegativeFloat = CONVERGENCE_GTOL_ABS
+    convergence_ftol_rel: NonNegativeFloat = CONVERGENCE_FTOL_REL
+    max_line_search_trials: PositiveInt = MAX_LINE_SEARCH_TRIALS
+    min_step_for_line_search: NonNegativeFloat = MIN_LINE_SEARCH_STEPS
+    max_step_for_line_search: NonNegativeFloat = MAX_LINE_SEARCH_STEPS
+
+    def _solve_internal_problem(
+        self, problem: InternalOptimizationProblem, x0: NDArray[np.float64]
+    ) -> InternalOptimizeResult:
+        optimizer = pye.L_BFGS(
+            numBasis=self.limited_memory_max_history,
+            maxIterations=self.stopping_maxiter,
+            armijoConstant=self.armijo_constant,
+            wolfe=self.wolfe_condition,
+            minGradientNorm=self.convergence_gtol_abs,
+            factr=self.convergence_ftol_rel,
+            maxLineSearchTrials=self.max_line_search_trials,
+            minStep=self.min_step_for_line_search,
+            maxStep=self.max_step_for_line_search,
+        )
+
+        def objective_function(
+            x: NDArray[np.float64], grad: NDArray[np.float64]
+        ) -> np.float64:
+            grad[:] = problem.jac(x)
+            return np.float64(problem.fun(x))
```

Comment:
Since both fun and jac are needed, it would be better to use `problem.fun_and_jac`. Then the `InternlOptimizationProblem` can decide which functions to call internally to get there. This is more efficient if a `fun_and_jac` was provided by the user or if we need to calculate numerical derivatives.